tls/s2n_connection.c

```diff
@@ -771,10 +771,12 @@ int s2n_connection_get_client_cert_chain(struct s2n_connection *conn, uint8_t **
     POSIX_ENSURE_REF(conn);
     POSIX_ENSURE_REF(der_cert_chain_out);
     POSIX_ENSURE_REF(cert_chain_len);
-    POSIX_ENSURE_REF(conn->secure.client_cert_chain.data);
+    POSIX_ENSURE_REF(conn->secure.peer_cert_chain.data);
 
-    *der_cert_chain_out = conn->secure.client_cert_chain.data;
-    *cert_chain_len = conn->secure.client_cert_chain.size;
+    POSIX_ENSURE(conn->mode == S2N_SERVER, S2N_ERR_INVALID_STATE);
```

Comment:
Use S2N_ERR_CLIENT if you throw an error here.

But wasn't the previous behavior to return no data if called from the server, not to error? Could this potentially break customers, and is it worth changing that behavior?

Reply:
The right behavior is to return the _validated_ cert chain. The older API had a sharp edge and returned full cert chain irrespective of whether x509 validation was skipped or not. Current customers who do not skip validation checks and expect the full cert chain will be affected, as we are now returning only the _validated_ cert chain (which is the correct behavior being fixed).